<a href="https://colab.research.google.com/github/yuraskass/lab1/blob/main/%D0%9A%D0%BE%D0%BF%D0%B8%D1%8F_%D0%B1%D0%BB%D0%BE%D0%BA%D0%BD%D0%BE%D1%82%D0%B0_%22%D0%97%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5_1_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Задание по практике 1

##Загрузка датасета
В файле classification datasets расположены ссылки на датасеты. Выберите один в соответствии со своим вариантом, скачайте и прочитайте файл

**Не забудьте загрузить нужные библиотеки**

Например: pip3 install pandas

In [2]:
import pandas as pd

In [4]:
df = pd.read_csv('/content/world-data-2023.csv')
df

Country Density\n(P/Km2) Abbreviation Agricultural Land( %)  \
0    Afghanistan               60           AF                58.10%   
1        Albania              105           AL                43.10%   
2        Algeria               18           DZ                17.40%   
3        Andorra              164           AD                40.00%   
4         Angola               26           AO                47.50%   
..           ...              ...          ...                   ...   
190    Venezuela               32           VE                24.50%   
191      Vietnam              314           VN                39.30%   
192        Yemen               56           YE                44.60%   
193       Zambia               25           ZM                32.10%   
194     Zimbabwe               38           ZW                41.90%   

    Land Area(Km2) Armed Forces size  Birth Rate  Calling Code  \
0          652,230           323,000       32.49          93.0   
1           28,748             9,000       11.78         355.0   
2        2,381,741           317,000       24.28         213.0   
3              468               NaN        7.20         376.0   
4        1,246,700           117,000       40.73         244.0   
..             ...               ...         ...           ...   
190        912,050           343,000       17.88          58.0   
191        331,210           522,000       16.75          84.0   
192        527,968            40,000       30.45         967.0   
193        752,618            16,000       36.19         260.0   
194        390,757            51,000       30.68         263.0   

    Capital/Major City Co2-Emissions  ... Out of pocket health expenditure  \
0                Kabul         8,672  ...                           78.40%   
1               Tirana         4,536  ...                           56.90%   
2              Algiers       150,006  ...                           28.10%   
3     Andorra la Vella           469  ...                           36.40%   
4               Luanda        34,693  ...                           33.40%   
..                 ...           ...  ...                              ...   
190            Caracas       164,175  ...                           45.80%   
191              Hanoi       192,668  ...                           43.50%   
192              Sanaa        10,609  ...                           81.00%   
193             Lusaka         5,141  ...                           27.50%   
194             Harare        10,983  ...                           25.80%   

    Physicians per thousand  Population  \
0                      0.28  38,041,754   
1                      1.20   2,854,191   
2                      1.72  43,053,054   
3                      3.33      77,142   
4                      0.21  31,825,295   
..                      ...         ...   
190                    1.92  28,515,829   
191                    0.82  96,462,106   
192                    0.31  29,161,922   
193                    1.19  17,861,030   
194                    0.21  14,645,468   

     Population: Labor force participation (%) Tax revenue (%) Total tax rate  \
0                                       48.90%           9.30%         71.40%   
1                                       55.70%          18.60%         36.60%   
2                                       41.20%          37.20%         66.10%   
3                                          NaN             NaN            NaN   
4                                       77.50%           9.20%         49.10%   
..                                         ...             ...            ...   
190                                     59.70%             NaN         73.30%   
191                                     77.40%          19.10%         37.60%   
192                                     38.00%             NaN         26.60%   
193                                     74.60%          16.20%         15.60%   
194   

##Первое знакомство с датасетом
В следующих ячейках выведите:


1.   Колонки датасета
2.   Первые 5 значений
3.   Последние 5 значений
4.   Статистику по числовым колонкам



In [5]:
df.columns

Index(['Country', 'Density\n(P/Km2)', 'Abbreviation', 'Agricultural Land( %)',
       'Land Area(Km2)', 'Armed Forces size', 'Birth Rate', 'Calling Code',
       'Capital/Major City', 'Co2-Emissions', 'CPI', 'CPI Change (%)',
       'Currency-Code', 'Fertility Rate', 'Forested Area (%)',
       'Gasoline Price', 'GDP', 'Gross primary education enrollment (%)',
       'Gross tertiary education enrollment (%)', 'Infant mortality',
       'Largest city', 'Life expectancy', 'Maternal mortality ratio',
       'Minimum wage', 'Official language', 'Out of pocket health expenditure',
       'Physicians per thousand', 'Population',
       'Population: Labor force participation (%)', 'Tax revenue (%)',
       'Total tax rate', 'Unemployment rate', 'Urban_population', 'Latitude',
       'Longitude'],
      dtype='object')

In [6]:
df.head(5)


Country Density\n(P/Km2) Abbreviation Agricultural Land( %)  \
0  Afghanistan               60           AF                58.10%   
1      Albania              105           AL                43.10%   
2      Algeria               18           DZ                17.40%   
3      Andorra              164           AD                40.00%   
4       Angola               26           AO                47.50%   

  Land Area(Km2) Armed Forces size  Birth Rate  Calling Code  \
0        652,230           323,000       32.49          93.0   
1         28,748             9,000       11.78         355.0   
2      2,381,741           317,000       24.28         213.0   
3            468               NaN        7.20         376.0   
4      1,246,700           117,000       40.73         244.0   

  Capital/Major City Co2-Emissions  ... Out of pocket health expenditure  \
0              Kabul         8,672  ...                           78.40%   
1             Tirana         4,536  ...                           56.90%   
2            Algiers       150,006  ...                           28.10%   
3   Andorra la Vella           469  ...                           36.40%   
4             Luanda        34,693  ...                           33.40%   

  Physicians per thousand  Population  \
0                    0.28  38,041,754   
1                    1.20   2,854,191   
2                    1.72  43,053,054   
3                    3.33      77,142   
4                    0.21  31,825,295   

   Population: Labor force participation (%) Tax revenue (%) Total tax rate  \
0                                     48.90%           9.30%         71.40%   
1                                     55.70%          18.60%         36.60%   
2                                     41.20%          37.20%         66.10%   
3                                        NaN             NaN            NaN   
4                                     77.50%           9.20%         49.10%   

  Unemployment rate Urban_population   Latitude  Longitude  
0            11.12%        9,797,273  33.939110  67.709953  
1            12.33%        1,747,593  41.153332  20.168331  
2            11.70%       31,510,100  28.033886   1.659626  
3               NaN           67,873  42.506285   1.521801  
4             6.89%       21,061,025 -11.202692  17.873887  

[5 rows x 35 columns]

In [7]:
df.tail(5)


Country Density\n(P/Km2) Abbreviation Agricultural Land( %)  \
190  Venezuela               32           VE                24.50%   
191    Vietnam              314           VN                39.30%   
192      Yemen               56           YE                44.60%   
193     Zambia               25           ZM                32.10%   
194   Zimbabwe               38           ZW                41.90%   

    Land Area(Km2) Armed Forces size  Birth Rate  Calling Code  \
190        912,050           343,000       17.88          58.0   
191        331,210           522,000       16.75          84.0   
192        527,968            40,000       30.45         967.0   
193        752,618            16,000       36.19         260.0   
194        390,757            51,000       30.68         263.0   

    Capital/Major City Co2-Emissions  ... Out of pocket health expenditure  \
190            Caracas       164,175  ...                           45.80%   
191              Hanoi       192,668  ...                           43.50%   
192              Sanaa        10,609  ...                           81.00%   
193             Lusaka         5,141  ...                           27.50%   
194             Harare        10,983  ...                           25.80%   

    Physicians per thousand  Population  \
190                    1.92  28,515,829   
191                    0.82  96,462,106   
192                    0.31  29,161,922   
193                    1.19  17,861,030   
194                    0.21  14,645,468   

     Population: Labor force participation (%) Tax revenue (%) Total tax rate  \
190                                     59.70%             NaN         73.30%   
191                                     77.40%          19.10%         37.60%   
192                                     38.00%             NaN         26.60%   
193                                     74.60%          16.20%         15.60%   
194                                     83.10%          20.70%         31.60%   

    Unemployment rate Urban_population   Latitude   Longitude  
190             8.80%       25,162,368   6.423750  -66.589730  
191             2.01%       35,332,140  14.058324  108.277199  
192            12.91%       10,869,523  15.552727   48.516388  
193            11.43%        7,871,713 -13.133897   27.849332  
194             4.95%        4,717,305 -19.015438   29.154857  

[5 rows x 35 columns]

In [8]:
df.describe()


Birth Rate  Calling Code  Fertility Rate  Infant mortality  \
count  189.000000    194.000000      188.000000        189.000000   
mean    20.214974    360.546392        2.698138         21.332804   
std      9.945774    323.236419        1.282267         19.548058   
min      5.900000      1.000000        0.980000          1.400000   
25%     11.300000     82.500000        1.705000          6.000000   
50%     17.950000    255.500000        2.245000         14.000000   
75%     28.750000    506.750000        3.597500         32.700000   
max     46.080000   1876.000000        6.910000         84.500000   

       Life expectancy  Maternal mortality ratio  Physicians per thousand  \
count       187.000000                181.000000               188.000000   
mean         72.279679                160.392265                 1.839840   
std           7.483661                233.502024                 1.684261   
min          52.800000                  2.000000                 0.010000   
25%          67.000000                 13.000000                 0.332500   
50%          73.200000                 53.000000                 1.460000   
75%          77.500000                186.000000                 2.935000   
max          85.400000               1150.000000                 8.420000   

         Latitude   Longitude  
count  194.000000  194.000000  
mean    19.092351   20.232434  
std     23.961779   66.716110  
min    -40.900557 -175.198242  
25%      4.544175   -7.941496  
50%     17.273849   20.972652  
75%     40.124603   48.281523  
max     64.963051  178.065032

**Ответьте на следующие вопросы:**


1.   Какие средние значения в каждой числовой колонке вашего датасета?
2.   Минимальные и максимальные значения в каждой числовой колонке вашего датасета





1.ср знач 20.214974	360.546392	2.698138	21.332804	72.279679	160.392265	1.839840	19.092351	20.232434
2.   мин 5.900000	1.000000	0.980000	1.400000	52.800000	2.000000	0.010000	-40.900557	-175.198242
макс 46.080000	1876.000000	6.910000	84.500000	85.400000	1150.000000	8.420000	64.963051	178.065032



##Типизация
Присвойте соответствующим колонкам один из типов (category, float, int)

In [41]:
df.Country	 = df.Country.astype("category")
df.Latitude	 = df.Latitude.astype("float64")
df.Longitude = df.Longitude.astype("float64")

df.info()


AttributeError: ignored

Какие признаки вы можете выделить в своем датасете (категориальные, числовые и др., **минимум 3 разных**, помните что можно также выделить признаки более точно, например числовые - порядковые, количественные и т.д.)?

##Фильтрация
Офильтруйте 3 колонки вашего датасета и выведите результат

In [31]:
df.loc[df['Longitude']<200]

Country Density\n(P/Km2) Abbreviation Agricultural Land( %)  \
0    Afghanistan               60           AF                58.10%   
1        Albania              105           AL                43.10%   
2        Algeria               18           DZ                17.40%   
3        Andorra              164           AD                40.00%   
4         Angola               26           AO                47.50%   
..           ...              ...          ...                   ...   
190    Venezuela               32           VE                24.50%   
191      Vietnam              314           VN                39.30%   
192        Yemen               56           YE                44.60%   
193       Zambia               25           ZM                32.10%   
194     Zimbabwe               38           ZW                41.90%   

    Land Area(Km2) Armed Forces size Birth Rate  Calling Code  \
0          652,230           323,000      32.49          93.0   
1           28,748             9,000      11.78         355.0   
2        2,381,741           317,000      24.28         213.0   
3              468               NaN       7.20         376.0   
4        1,246,700           117,000      40.73         244.0   
..             ...               ...        ...           ...   
190        912,050           343,000      17.88          58.0   
191        331,210           522,000      16.75          84.0   
192        527,968            40,000      30.45         967.0   
193        752,618            16,000      36.19         260.0   
194        390,757            51,000      30.68         263.0   

    Capital/Major City Co2-Emissions  ... Out of pocket health expenditure  \
0                Kabul         8,672  ...                           78.40%   
1               Tirana         4,536  ...                           56.90%   
2              Algiers       150,006  ...                           28.10%   
3     Andorra la Vella           469  ...                           36.40%   
4               Luanda        34,693  ...                           33.40%   
..                 ...           ...  ...                              ...   
190            Caracas       164,175  ...                           45.80%   
191              Hanoi       192,668  ...                           43.50%   
192              Sanaa        10,609  ...                           81.00%   
193             Lusaka         5,141  ...                           27.50%   
194             Harare        10,983  ...                           25.80%   

    Physicians per thousand  Population  \
0                      0.28  38,041,754   
1                      1.20   2,854,191   
2                      1.72  43,053,054   
3                      3.33      77,142   
4                      0.21  31,825,295   
..                      ...         ...   
190                    1.92  28,515,829   
191                    0.82  96,462,106   
192                    0.31  29,161,922   
193                    1.19  17,861,030   
194                    0.21  14,645,468   

     Population: Labor force participation (%) Tax revenue (%) Total tax rate  \
0                                       48.90%           9.30%         71.40%   
1                                       55.70%          18.60%         36.60%   
2                                       41.20%          37.20%         66.10%   
3                                          NaN             NaN            NaN   
4                                       77.50%           9.20%         49.10%   
..                                         ...             ...            ...   
190                                     59.70%             NaN         73.30%   
191                                     77.40%          19.10%         37.60%   
192                                     38.00%             NaN         26.60%   
193                                     74.60%          16.20%         15.60%   
194               

In [36]:
df.loc[df['Latitude']<100]

Country Density\n(P/Km2) Abbreviation Agricultural Land( %)  \
0    Afghanistan               60           AF                58.10%   
1        Albania              105           AL                43.10%   
2        Algeria               18           DZ                17.40%   
3        Andorra              164           AD                40.00%   
4         Angola               26           AO                47.50%   
..           ...              ...          ...                   ...   
190    Venezuela               32           VE                24.50%   
191      Vietnam              314           VN                39.30%   
192        Yemen               56           YE                44.60%   
193       Zambia               25           ZM                32.10%   
194     Zimbabwe               38           ZW                41.90%   

    Land Area(Km2) Armed Forces size Birth Rate  Calling Code  \
0          652,230           323,000      32.49          93.0   
1           28,748             9,000      11.78         355.0   
2        2,381,741           317,000      24.28         213.0   
3              468               NaN       7.20         376.0   
4        1,246,700           117,000      40.73         244.0   
..             ...               ...        ...           ...   
190        912,050           343,000      17.88          58.0   
191        331,210           522,000      16.75          84.0   
192        527,968            40,000      30.45         967.0   
193        752,618            16,000      36.19         260.0   
194        390,757            51,000      30.68         263.0   

    Capital/Major City Co2-Emissions  ... Out of pocket health expenditure  \
0                Kabul         8,672  ...                           78.40%   
1               Tirana         4,536  ...                           56.90%   
2              Algiers       150,006  ...                           28.10%   
3     Andorra la Vella           469  ...                           36.40%   
4               Luanda        34,693  ...                           33.40%   
..                 ...           ...  ...                              ...   
190            Caracas       164,175  ...                           45.80%   
191              Hanoi       192,668  ...                           43.50%   
192              Sanaa        10,609  ...                           81.00%   
193             Lusaka         5,141  ...                           27.50%   
194             Harare        10,983  ...                           25.80%   

    Physicians per thousand  Population  \
0                      0.28  38,041,754   
1                      1.20   2,854,191   
2                      1.72  43,053,054   
3                      3.33      77,142   
4                      0.21  31,825,295   
..                      ...         ...   
190                    1.92  28,515,829   
191                    0.82  96,462,106   
192                    0.31  29,161,922   
193                    1.19  17,861,030   
194                    0.21  14,645,468   

     Population: Labor force participation (%) Tax revenue (%) Total tax rate  \
0                                       48.90%           9.30%         71.40%   
1                                       55.70%          18.60%         36.60%   
2                                       41.20%          37.20%         66.10%   
3                                          NaN             NaN            NaN   
4                                       77.50%           9.20%         49.10%   
..                                         ...             ...            ...   
190                                     59.70%             NaN         73.30%   
191                                     77.40%          19.10%         37.60%   
192                                     38.00%             NaN         26.60%   
193                                     74.60%          16.20%         15.60%   
194               

In [42]:
df.loc[df['Country']=="India"]

Country Density\n(P/Km2) Abbreviation Agricultural Land( %) Land Area(Km2)  \
77   India              464           IN                60.40%      3,287,263   

   Armed Forces size Birth Rate  Calling Code Capital/Major City  \
77         3,031,000      17.86          91.0          New Delhi   

   Co2-Emissions  ... Out of pocket health expenditure  \
77     2,407,672  ...                           65.10%   

   Physicians per thousand     Population  \
77                    0.86  1,366,417,754   

    Population: Labor force participation (%) Tax revenue (%) Total tax rate  \
77                                     49.30%          11.20%         49.70%   

   Unemployment rate Urban_population   Latitude  Longitude  
77             5.36%      471,031,528  20.593684   78.96288  

[1 rows x 35 columns]

##Группировка
Сгруппируйте датасет по какому-либо категориальному признаку и выведите среднее значение численных колонок (см. метод mean в pandas)

In [43]:
df.groupby('Longitude').mean()

<ipython-input-43-36df232993fa>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df.groupby('Longitude').mean()


Calling Code  Fertility Rate  Infant mortality  Life expectancy  \
Longitude                                                                      
-175.198242         676.0            3.56              13.4             70.8   
-172.104629         685.0            3.88              13.6             73.2   
-157.376832         686.0            3.57              41.2             68.1   
-106.346771           1.0            1.50               4.3             81.9   
-102.552784          52.0            2.13              11.0             75.0   
...                   ...             ...               ...              ...   
 166.959158         678.0            3.78              22.3             70.3   
 171.184478         692.0            4.05              27.4             65.2   
 174.885971          64.0            1.71               4.7             81.9   
 177.649330         688.0             NaN              20.6              NaN   
 178.065032         679.0            2.77              21.6             67.3   

             Maternal mortality ratio  Physicians per thousand   Latitude  
Longitude                                                                  
-175.198242                      52.0                     0.52 -21.178986  
-172.104629                      43.0                     0.34 -13.759029  
-157.376832                      92.0                     0.20   1.836898  
-106.346771                      10.0                     2.61  56.130366  
-102.552784                      33.0                     2.38  23.634501  
...                               ...                      ...        ...  
 166.959158                      72.0                     0.17 -15.376706  
 171.184478                       NaN                     0.42   7.131474  
 174.885971                       9.0                     3.59 -40.900557  
 177.649330                       NaN                     0.92  -7.109535  
 178.065032                      34.0                     0.84 -17.713371  

[194 rows x 7 columns]

##Задание для гениев *
Агрегируйте числовой признак по категориальному признаку с использованием метода **без использования** библиотеки pandas, и вычислите медиану и среднее значение для каждой категории.

In [ ]:
#Агрегация